### ISSUE TITLE
Microsoft.Data.Sqlite.SqliteException: 'SQLite Error 5: 'database is locked'.'

### ISSUE THREAD

**Issue by Tuytje** [Mimetis/Dotmim.Sync#576]
Recently I have been getting reports about a locked (sqlite) database, but only when sync is enabled. I can simulate it in debug and when I comment out the syncing part, it does not lock up anymore. 

This wasn't the case some time ago, no db locking was done. I'm not sure what exactly changed & any pointers as to why the database gets locked by the sync process would be greatly appreciated. 
Or maybe it did lock the database before as well, but the sync was faster so it wasn't an issue. I did notice the sync taking some time (regular sync, not the initial sync)

The error I'm getting is "SQLite Error 5: 'database is locked'"

Sync Log:

[0:] 17:47:13.595 -> --==>> Sync START. (17:47:13.595)
[0:] 17:47:16.047 -> ScopeLoading:	Getting Server Scope. Scope Name:DefaultScope.
[0:] 17:47:27.044 -> ChangesSelecting:	[_bg_BoardGame] [Total] Upserts:1. Deletes:0. Total:1.
[0:] 17:47:27.063 -> ChangesSelecting:	[_bg_BoardGameLink] [Total] Upserts:5. Deletes:0. Total:5.
[0:] 17:47:27.089 -> ChangesSelecting:	[_bg_Media] [Total] Upserts:1. Deletes:0. Total:1.
[0:] 17:47:27.192 -> ChangesSelecting:	[Total] Upserts:7. Deletes:0. Total:7
[0:] 17:47:27.580 -> ChangesApplying:	Sending Batch Changes. Batches: (1/1). Rows: (7/7). Waiting Server Response...
[0:] 17:47:30.484 -> ChangesSelecting:	Downloading Batches. Scope Name:DefaultScope. Batches Count:1. Rows Count:12855
[0:] 17:47:36.034 -> ChangesSelecting:	Downloaded Batch Changes. (1/1). Rows:(12855/12855).
[0:] 17:47:36.313 -> ChangesSelecting:	Snapshot Downloaded. Batches Count: 1. Total Rows: 12855. Duration: 00:00:22
[0:] 17:47:43.851 -> ChangesApplying:	[_bg_BoardGame] [Modified] Applied:(0) Total:(0/12855).
[0:] 17:48:12.992 -> ChangesApplying:	[_bg_BoardGameLink] [Modified] Applied:(0) Total:(0/12855).
[0:] 17:48:19.676 -> ChangesApplying:	[_bg_Collection] [Modified] Applied:(0) Total:(0/12855).
[0:] 17:48:39.079 -> ChangesApplying:	[_bg_Media] [Modified] Applied:(0) Total:(0/12855).
**Microsoft.Data.Sqlite.SqliteException:** 'SQLite Error 5: 'database is locked'.'

As can be seen, the initial part (fetching it from the server) is pretty fast, but applying those changes locally (I think it is that) is taking a long time. Ths is very strange since those tables don't contain a lot of records. for example the _bg_collection table contains at most 7 records. How can this take that long?

What does the number (0/12855) at the end mean? 

Any pointers would be appreciated

**Comment by Mimetis** [comment]
you may have a connection opened somewhere else in your code, that prevents DMS to open a new one

**Comment by Tuytje** [comment]
@Mimetis I'm sorry if it wasn't clear, the code itself works perfectly when sync is not enabled. It seems that the sync itself locks the database. Is that intended behaviour?

What would the reason be that ChangesApplying takes so long? If there any way that I can speed that up?

**Comment by Mimetis** [comment]
Well, if you have an open connection, before launching sync, then the sync may fail at some point.
I guess the time is too long, because Sqlite is waiting for the lock to be released, then timeout
Are you sure all the connections are closed; before trying a sync ?

**Comment by Tuytje** [comment]
@Mimetis thank you for the swift reply. 

All my connections are closed. When disabeling the sync, the app works perfectly. It seems that the sync library is holding the lock having my code fail. 

The syncing happens in a background thread, since we don't want it to interfere with the normal flow of the app. This has work perfectly. The syncing itself has worked as intended for a couple of months, until now. The facts that it now takes so long to apply the changes, for a table with 8 rows might have something to do with that. But what could be the reason? 
There used to be now interference with the application code, gave me the impression that no table locks were happening for the sync library. Am I correct in this?

**Comment by Mimetis** [comment]
That's what I mean
Nothing has changed in the `DMS` package (I guess you are working with the same version)
Has your code changed at some point, since the first locks appeared?

IF you are able to reproduce the bug, try this:
Instead of launching `DMS`, open a new `SqliteConnection`, in the same background thread, and make an update, and see if the update is successful.

IF you are not able to reproduce the bug, try to sync on start, before making any open connection, and see if it's working.

Try to see if you are launching a sync, when you have a connection already opened on sqlite.

If you have a lock, it's for sure a transaction not committed, that has been opened before you launch the DMS sync

**Comment by Tuytje** [comment]
This is the thing, nothing has changed on that end as well. A sync is started when the app starts, but that sync takes a lot longer than it used to take and that has as a result that everything fails since it locks the database. I have no idea if it used to lock the database, but the sync itself was done pretty fast, so it wouldn't have even mattered. What reason could there be to have the sync take so long?

As previously pasted: 
[0:] 17:48:19.676 -> ChangesApplying: [_bg_Collection] [Modified] Applied:(0) Total:(0/12855).
[0:] 17:48:39.079

Am i correct in thinking that 'ChangesApplying' is doing the local sync? If says (0), so that means nothing was to be applied? What is the 'Total' behind that? That table has 8 records, still it is taking 20 seconds, which seems strange

**Comment by Mimetis** [comment]
Ok , indeed, it's strange.

Is it possible for you, to create a blank xamarin application, with just a push button and the sync code behind it ?
Using an empty sqlite database, and then a second shot, using your existing client database ?
Let the server side, as is, it seems the problem is not coming from it.

If you are still facing the issue, then the problem is coming from `DMS`
If that's the case, we can try to debug it, and find what is struggling here

**Comment by Tuytje** [comment]
At this time I cannot because of a shortage of time :(

I'm pretty sure that the problem is specific to this app, somehow I have a feeling that the fact that I'm running the sync for around 3 months has something to do with it. 

Could the sync be getting slower? I have no snapshots, but since the number of records in the tables is low (each table has a couple of dozen records, except one who could be a couple of hundred) it didn't seem an issue to me (and space on the server side is limited)

What are those number: Applied:(0) Total:(0/12855) ?

Is there a way for DMS to take record locks, not table locks?

Since the issue only arises when users edit / add data, and they usually read data I could build in a sync cancel. I have looked around, but I can't seem to find a way to cancel a sync in progress. Did I miss something?

These syncs are after a full sync (new install, pulling in all data). The full sync works perfectly since a popup blocks all user input in that case. After that initial sync, there should be no changes coming in. Which is why it is very strange that a 'Normal' sync afterwards takes so much time.

**Comment by Tuytje** [comment]
Some extra info:
I just noticed that all rows are always in conflict (summary after sync):
	Total changes  uploaded: 0
	Total changes  downloaded: 13314 
	Total changes  applied: 0 
	Total resolved conflicts: 13314
	Total duration :00.00:08.822 

There are no 13.000 rows, so no idea where that number comes from. How could they all be in conflict since the just got downloaded from the server?

**Comment by Mimetis** [comment]
Oh, so you have 13K rows in conflicts.
That's explains a lot indeed.
If you have conflicts here, then the rows may exists on the client, as well, right ?
If so, you need to understand why you have conflicts on the client, 

you can use the `localOrchestrator.OnApplyChangesFailed` interceptor and see what kind of conflict you have on the client, with the args
Can you test that ?

**Comment by Tuytje** [comment]
It does not get hit, I have added it like so:

            clientOrchestrator.OnApplyChangesFailed(action => 
            {
                  Helper.LogDebug($"OnApplyChangesFailed - Message: {action.Message}. Conflict: {action.Conflict.ToString()}");
            });

The clientOrchestrator was created as following:

            var handler = new HttpClientHandler { AutomaticDecompression = DecompressionMethods.GZip };
            
            var client = new HttpClient(handler) { Timeout = TimeSpan.FromMinutes(20) };

            var clientOrchestrator = new WebClientOrchestrator(SYNC_URL, client: client);

The output is still:

14:25:08.273 -> --==>> Sync SUCCESS (14:25:08.273): Synchronization done. 
	Total changes  uploaded: 0
	Total changes  downloaded: 13315 
	Total changes  applied: 0 
	Total resolved conflicts: 13315
	Total duration :00.00:07.720 

I'm running the code on windows now, easier debugging, hence the faster processing. But the same problem remains. Everytime I sync, it will give me those numbers, even though I just sync'd.

Any thoughts?

**Comment by Mimetis** [comment]
Ok, now, still on the client test with this interceptor :
``` cs
localOrchestrator.OnTableChangesBatchApplying()
```
You will have 
- `changes` : A datatable containins a table batch to apply
- `state`: An enumeration with the type of apply (insert, delete, update)

That way, you will see what are these 13k rows (in my opinion, ghost deletes)

That being said, I don't understand why you are not hitting the ``localOrchestrator.OnApplyChangesFailed()` interceptor...

Well, firstly, let's see what you have in your batches :)

**Comment by Tuytje** [comment]
I've tried with that interceptor, but it is also not getting hit.

I have a couple now, and only the last 2 are getting hit. I have been expecting them, but I see no data only a row count.

            clientOrchestrator.OnApplyChangesFailed(action => 
            {
                Helper.LogDebug($"OnApplyChangesFailed - Message: {action.Message}. Conflict: {action.Conflict.ToString()}");
            });

            clientOrchestrator.OnTableChangesBatchApplying(action =>
            {
                Helper.LogDebug($"OnTableChangesBatchApplying - Message: {action.Message}. Conflict: {action.Changes.ToString()}");
            });

            clientOrchestrator.OnTableChangesApplying(action =>
            {
                Helper.LogDebug($"OnTableChangesApplying - Message: {action.Message}. Conflict: {action.Context.ToString()}");
            });

            clientOrchestrator.OnHttpBatchesDownloadingArgs(action =>
            {
                Helper.LogDebug($"OnHttpBatchesDownloadingArgs - Message: {action.Message}. Conflict: {action.Context.ToString()}");
            });

            clientOrchestrator.OnHttpGettingChangesResponse(action =>
            {
                Helper.LogDebug($"OnHttpGettingChangesResponse - Message: {action.Message}. Conflict: {action.Context.ToString()}");
            });

Any other interceptor that I could try to get more info?

**Comment by Tuytje** [comment]
I found something on OnHttpBatchesDownloadedArgs and it seems your hunch was correct, they are mostly deletes ... How to proceed with them?

**Comment by Mimetis** [comment]
That's not normal that interceptors (excepting the `OnHttpBatchesDownloaded`) are not working.

Here are my thought:
You are probably using a filter for all your clients.
Filters are working great, except for `Deletes`.
All `Deletes` from _everywhere_ are sent to **all** clients.
Why ? Because each delete in tracking tables, are saved with the primary key, but without any consideration to filter.

Once these deletes are sent, they will not be sent again, of course.
That's why I don't understand why you have these deletes sent, after a successful initialization....

You can purge the Server metadatas, calling the `await remOrchestrator.DeleteMetadatasAsync();` method.
This is something you should do every two or three weeks, I guess.

But once again, there is something wrong here:
- no interceptors catched
- deletes downloaded after an init phase, that should not happened, since your last sync timestamp should be > to all deletes

**Comment by Tuytje** [comment]
I have done this, but there is no resolvement. It found 149 histories, so at least something is picked up. They all sync'd in the last 2 weeks, which results in a minTimestamp of 9983. When using this timestamp to delete the metadata, it has no results in the resulting DatabaseMetadatasCleaned object ...

At this time the system is at almost 18.000 'conflicts'.

I have changed the retention time of sql server change tracking from 90 days to 30 days some time ago, not sure if it could be the cause, but I thought to mention it. Could it be the cause?

Syncing changes still works since I can make a change, or add new entries and when a sync is started and no other database action is done until the sync is done, then the change can be found on the server.

An initial sync can also still be done, when it does all data is pulled in onto the client.

Any thoughts?

**Comment by Tuytje** [comment]
@Mimetis I have done some more testing and the strange thing is that the system works. If a new entry is made, it is uploaded and other clients are able to downloading those changes. Those changes only download once as well. It seems that the deletes are the ones who are 'stuck'. Any idea why this might be? 

Is there any solution for having only 'your' deletes? I understand the problem, but for larger systems this might become a problem. 

Is there any way for me to purge the system of those 'stuck' deletes? I have tried with the DeleteMetadata method, but it seems that one only find the not stuck items. The deletes don't show up.

Something that strikes me as odd: I have tracking tables on the server, but I used SQL Server Change Tracking. As I understand it, this should mean that I shouldn't have any tracking tables, but I do and they have recent data. I double checked, change tracking is enabled on the database and I'm using the dotmim.sync.sqlserver.changetracking package on the server. 

Worst case scenario: Can I 'reset' the system? With that I mean lose all changes and continue with the current data in the database.

**Comment by Mimetis** [comment]
Since you are using the Change Tracking mechanism, the purge is down by SQL Server itself.

My advice on using `await remOrchestrator.DeleteMetadatasAsync();` is only true when you are using tracking tables with `SqlSyncProvider`.

If you have tracking tables on your server, it's maybe because you've made some tests with a simple `SqlSyncProvider` before using `SqlChangeTrackingSyncProvider` ?

By the way, you can delete them (be careful to be sure you are not using any `SqlSyncProvider` )

If you have recent data in them, it's probably because you have old triggers that are not necessary anymore.
With Change Tracking, you should not have any triggers in your tables, since it's handled by the system

**Comment by Tuytje** [comment]
I managed to get the system working again and purge the (by now) 18.000 deletes by changing to regular tracking. Something goes wrong when I try to use 'Dotmim.Sync.SqlSyncChangeTrackingProvider'. 

No errors, the system seems to work, but it creates the triggers & tables and when requesting the web server properties it doesn't show me 'useChangeTracking: enabled', although the database did. It even marked the correct tables as 'tracked', but the SqlSyncChangeTrackingProvider dit not use the change tracking. It is as if the SqlSyncChangeTrackingProvider acts just like the regular SqlSyncProvider.

I have now changed to using SqlSyncProvider & disabling change tracking on sql server and all is good. I have a couple of older backups and it seems that for a backup from 4/06 I had no tracking tables and a backup from 1/07 is showing me tracking tables. Not sure what happened there and why it is not working anymore. I would prefer to use the sql server change tracking provider of course since it auto cleans. Any pointers would be appreciated.

**Comment by Mimetis** [comment]
Ok, let's have an eye on this in the feature.
I will close the issue since you found a way to resolve it.
Do not hesitate to reopen it (or ask to) if you think we should continue to troubleshoot it

**Comment by VagueGit** [comment]
We have recently experienced DMS reporting Database is locked in our test environment. All our connections are within using statements and are disposed before sync is called.

We use `DMS 0.8.0` with `maxDownladingDegreeOfParallelism: 4`. We sync on the UI thread.

The error only seems to occur when testing sync >100mb. The test database is 44 tables. SQLite on client, SQL Server with change tracking on server via web api.

I found that by breaking the large sync into multiple scopes and syncing one scope at a time, the issue went away. So we have an acceptable workaround.
